### DISCUSSION TITLE
Custom Resource Definitions and Cluster Upgrades

### DISCUSSION THREAD

**Discussion by atsai1220** [liqotech/liqo#2062] [Q&A] [answered]
Hello!

What is the expectation when one of the peered cluster's Kubernetes version is upgraded and there are deprecated APIs? What happens when one of the cluster has a CRD version upgraded as well? Thank you!

Best,

Andrew

**Comment by frisso** [comment]
@atsai1220 The short answer is that we're not in a 1.0 release yet, and some incompatibilities can occur when migrating to the most recent version. There are some companies that are considering buying support to address those cases, e.g., when Liqo is already in production (such as in our University).
For a more technical insight, I leave the floor to the tech people here.

**Comment by cheina97** [comment]
> What is the expectation when one of the peered cluster's Kubernetes version is upgraded and there are deprecated APIs?

The K8S API is quite stable, in case there are some incompatibilities with Liqo we will notice it and fix it. We guarantee the compatibility with the latest 3 major K8S releases (we added this detail in documentation today).

> What happens when one of the cluster has a CRD version upgraded as well? Thank you!

It is something we are planning to manage.
As @frisso said we are moving to a stable API in order to avoid changes in CRDs in the future. We are also evaluating the best strategy to manage CRDs changes in the future. At the moment we don't provide a method to upgrade CRDs and we suggest to reinstall Liqo.

**Comment by atsai1220** [comment]
Thank you for the responses! Does that mean any 3rd party CRDs and Kubernetes versions will need to be the same across peered clusters? What would the upgrade process look like for peered clusters? One at a time and hope for the best or make a new cluster and peer with an existing "cluster" of clusters? Still wrapping my head around Liqo and much appreciate the insight. Thank you!

**Comment by frisso** [comment, marked as answer]
> Thank you for the responses! Does that mean any 3rd party CRDs and Kubernetes versions will need to be the same across peered clusters?

No, this is not needed, Kubernetes versions can be different. However, given that the API is not yet stable, Liqo should be the same version on both clusters.
 
> What would the upgrade process look like for peered clusters? One at a time and hope for the best or make a new cluster and peer with an existing "cluster" of clusters? Still wrapping my head around Liqo and much appreciate the insight. Thank you!

Currently, given that we're not 1.0 for Liqo, "upgrading" is not guaranteed. It may work, but we cannot make promises. So, the safer way is to remove Liqo and install the new version. We know that this is really a pain, e.g., if you have a cluster already in production (such as in our University), hence the reason we can offer support for this operation, if needed.
Note that you don't have to create a new cluster from scratch; removing Liqo and installing the new version is enough.
